### PR TITLE
Include ISODate 📆 for constructorCall 👷‍♀️📞 and handle appropriately

### PR DESCRIPTION
**Pull request by nturinski**
Fixes https://github.com/microsoft/vscode-cosmosdb/issues/1096

The error was happening because ISODate was being handled the same as a Date object.  The string representation for a Date is `1975-07-23T04:10:00.000` as opposed to ISODate's `1975-07-23T04:10:00.000Z`.  So when Mongo inserted the date, it just assumed the timezone was whatever the computer's local timezone was set to.

Edit: Both `db.test.insert({ 'datetime': ISODate('1975-07-23T04:10:00') })` and `db.test.insert({ 'datetime': ISODate('1975-07-23T04:10:00Z') })` are valid timestrings.

### CODE REVIEW

File: src/mongo/MongoScrapbook.ts [hunk 1]
```diff
@@ -348,35 +348,55 @@ class FindMongoCommandsVisitor extends MongoVisitor<MongoCommand[]> {
...
 
-		let functionMap = { "ObjectId": this.objectIdToObject, "ISODate": this.dateToObject, "Date": this.dateToObject };
+		let functionMap = { "ObjectId": this.objectIdToObject, "ISODate": this.isodateToObject, "Date": this.dateToObject };
 		if (argumentContextArray.length > 1) {
 			let err: IParsedError = parseError(`Too many arguments. Expecting 0 or 1 argument(s) to ${constructorCall}`);
 			this.addErrorToCommand(err, ctx);
 			return {};
 		}
+
 		if (constructorCall.text in functionMap) {
```
Comment (EricJizbaMSFT) [inline src/mongo/MongoScrapbook.ts]: This `functionMap` thing is really weird to me. I feel like it'll make our lives easier if you just do a simple `switch` like this:

```typescript
const tokenText: string | undefined = argumentContextArray.length ? argumentContextArray[0].text : undefined;
switch (constructorCall.text) {
	case 'ObjectId':
		return this.objectIdToObject(ctx, tokenText);
	case 'ISODate':
		return this.dateToObject(ctx, tokenText, true /* isIsoDate */);
	case 'Date':
		return this.dateToObject(ctx, tokenText);
	default:
		// ...
}
```

File: test/mongoGetCommand.test.ts [hunk 2]
```diff
@@ -459,14 +459,50 @@ suite("scrapbook parsing Tests", () => {
...
+                    "date": {
+                        "$date": "2018-05-01T00:00:00.000Z"
+                    },
+                    "name": "Stephen"
+                }
+            ]
+        });
+    });
+
+    test("ISODate 3", () => {
```
[the remark this answers is not in the document]
Comment (nturinski) [reply test/mongoGetCommand.test.ts]: image

File: src/mongo/MongoScrapbook.ts [hunk 3]
```diff
@@ -347,37 +347,68 @@ class FindMongoCommandsVisitor extends MongoVisitor<MongoCommand[]> {
...
+		}
+	}
+
+	private dateToObject(ctx: mongoParser.ArgumentContext | mongoParser.PropertyValueContext, tokenText?: string): { $date: Date | {} } {
+		let stringDate: Date | {} = this.tryToConstructDate(ctx, tokenText);
+		if (stringDate instanceof Date) {
+			return { $date: stringDate.toISOString() };
+		}
+
+		return { $date: stringDate };
```
Comment (EricJizbaMSFT) [inline src/mongo/MongoScrapbook.ts]: I think the return type of `dateToObject` should be `{ $date: string } | {}`. 

No need to add the "$date" stuff if it returned an empty object (aka in the error case below)

File: src/mongo/tree/MongoCollectionTreeItem.ts [hunk 4]
```diff
@@ -183,7 +183,9 @@ export class MongoCollectionTreeItem extends AzureParentTreeItem<IMongoTreeRoot>
 		if (!document) {
 			throw new Error("The insert command requires at least one argument");
 		}
-		const insertResult = await this.collection.insert(document);
+
+		// insert is deprecated, so use insertOne instead
+		const insertResult = await this.collection.insertOne(document);
```
Comment (EricJizbaMSFT) [inline src/mongo/tree/MongoCollectionTreeItem.ts]: Was this intentional? Seems unrelated

File: src/mongo/MongoScrapbook.ts [hunk 5]
```diff
@@ -347,37 +347,68 @@ class FindMongoCommandsVisitor extends MongoVisitor<MongoCommand[]> {
...
+				let unrecognizedNodeErr: IParsedError = parseError(`Unrecognized node type encountered. Could not parse ${constructorCall.text} as part of ${child.text}`);
+				this.addErrorToCommand(unrecognizedNodeErr, ctx);
+				return {};
+		}
+	}
+
+	private dateToObject(ctx: mongoParser.ArgumentContext | mongoParser.PropertyValueContext, tokenText?: string): { $date: Date | {} } {
+		let stringDate: Date | {} = this.tryToConstructDate(ctx, tokenText);
+		if (stringDate instanceof Date) {
+			return { $date: stringDate.toISOString() };
```
Comment (EricJizbaMSFT) [inline src/mongo/MongoScrapbook.ts]: Previously it did `toString` instead of `toISOString`. Honestly I'm not sure what's the expected behavior here, but I would err on the side of keeping it consistent with previous releases

File: src/mongo/tree/MongoCollectionTreeItem.ts (hunk 4 above)
Comment (nturinski) [reply src/mongo/tree/MongoCollectionTreeItem.ts]: It was just something I noticed when trying to resolve this issue (since I thought the bug might have stemmed from using an unsupported insert command).  I can remove it though

Comment (EricJizbaMSFT) [reply src/mongo/tree/MongoCollectionTreeItem.ts]: I'd prefer if you removed it just so I don't have to investigate that file haha. At the moment I'm not exactly sure what's going on there
